File: TexasProgram/Models/BroChatModels.swift
import Foundation

enum BroMessageType: String, Codable, Sendable {
    case text = "TEXT"
    case workoutResult = "WORKOUT_RESULT"
    case photo = "PHOTO"

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let raw = (try? container.decode(String.self)) ?? "TEXT"
        switch raw.uppercased() {
        case "TEXT": self = .text
        case "WORKOUT_RESULT": self = .workoutResult
        case "PHOTO": self = .photo
        default: self = .text
        }
    }
}

struct WorkoutSharePayload: Codable, Equatable, Hashable, Sendable {
    var exercise: String
    var weight: String
    var sets: String
    var reps: String
    var isPR: Bool
    var week: Int
    var day: Int
}

struct BroChatMessage: Codable, Identifiable, Equatable, Hashable, Sendable {
    let id: String
    let channelId: String
    let senderId: String
    let senderName: String
    let timestamp: Int64
    let text: String
    let type: BroMessageType
    var workoutPayload: WorkoutSharePayload?
    var photoBase64: String?

    init(
        id: String = UUID().uuidString,
        channelId: String,
        senderId: String,
        senderName: String,
        timestamp: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        text: String = "",
        type: BroMessageType = .text,
        workoutPayload: WorkoutSharePayload? = nil,
        photoBase64: String? = nil
    ) {
        self.id = id
        self.channelId = channelId
        self.senderId = senderId
        self.senderName = senderName
        self.timestamp = timestamp
        self.text = text
        self.type = type
        self.workoutPayload = workoutPayload
        self.photoBase64 = photoBase64
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        channelId = try container.decode(String.self, forKey: .channelId)
        senderId = try container.decode(String.self, forKey: .senderId)
        senderName = try container.decode(String.self, forKey: .senderName)
        timestamp = try container.decode(Int64.self, forKey: .timestamp)
        text = try container.decodeIfPresent(String.self, forKey: .text) ?? ""
        type = (try? container.decode(BroMessageType.self, forKey: .type)) ?? .text
        workoutPayload = try container.decodeIfPresent(WorkoutSharePayload.self, forKey: .workoutPayload)
        photoBase64 = try container.decodeIfPresent(String.self, forKey: .photoBase64)
    }

    private enum CodingKeys: String, CodingKey {
        case id, channelId, senderId, senderName, timestamp, text, type, workoutPayload, photoBase64
    }

    var timeFormatted: String {
        let date = Date(timeIntervalSince1970: Double(timestamp) / 1000.0)
        let formatter = DateFormatter()
        formatter.dateFormat = "HH:mm"
        return formatter.string(from: date)
    }
}

struct BroChatChannelPayload: Codable, Sendable {
    let channelId: String
    var messages: [BroChatMessage]
    var updatedAt: Int64
}

File: android/app/src/main/java/com/texasprogram/app/model/BroChatModels.kt
package com.texasprogram.app.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID

import kotlinx.serialization.KSerializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

object BroMessageTypeSerializer : KSerializer<BroMessageType> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("BroMessageType", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: BroMessageType) {
        encoder.encodeString(value.name)
    }

    override fun deserialize(decoder: Decoder): BroMessageType {
        val raw = try { decoder.decodeString().uppercase() } catch (_: Exception) { "TEXT" }
        return when (raw) {
            "TEXT" -> BroMessageType.TEXT
            "WORKOUT_RESULT" -> BroMessageType.WORKOUT_RESULT
            "PHOTO" -> BroMessageType.PHOTO
            else -> BroMessageType.TEXT
        }
    }
}

@Serializable(with = BroMessageTypeSerializer::class)
enum class BroMessageType {
    @SerialName("TEXT") TEXT,
    @SerialName("WORKOUT_RESULT") WORKOUT_RESULT,
    @SerialName("PHOTO") PHOTO
}

@Serializable
data class WorkoutSharePayload(
    val exercise: String,
    val weight: String,
    val sets: String,
    val reps: String,
    val isPR: Boolean = false,
    val week: Int = 1,
    val day: Int = 1
)

@Serializable
data class BroChatMessage(
    val id: String = UUID.randomUUID().toString(),
    val channelId: String,
    val senderId: String,
    val senderName: String,
    val timestamp: Long = System.currentTimeMillis(),
    val text: String = "",
    val type: BroMessageType = BroMessageType.TEXT,
    val workoutPayload: WorkoutSharePayload? = null,
    val photoBase64: String? = null
) {
    val timeFormatted: String
        get() {
            val date = Date(timestamp)
            val formatter = SimpleDateFormat("HH:mm", Locale.getDefault())
            return formatter.format(date)
        }
}

@Serializable
data class BroChatChannelPayload(
    val channelId: String,
    val messages: List<BroChatMessage> = emptyList(),
    val updatedAt: Long = System.currentTimeMillis()
)
